CustomGradientPicker: Refactor to TypeScript

chad1008 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Refactor `CustomGradientPicker` component to TypeScript

Part of https://github.com/WordPress/gutenberg/issues/35744

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The refactor to TypeScript has many benefits (auto-generated docs, static linting and error prevention, better IDE experience). See https://github.com/WordPress/gutenberg/issues/35744 for more details

## How?
Followed the steps in the [TypeScript migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2022-11-29T14%3A45%3A34Z#refactoring-a-component-to-typescript)

## Testing Instructions
- After applying the patch, run `npm ci`, as a new `DefinitelyTyped` package has been added for the `gradientParser` library this component relies on.
- Ensure all unit and e2e tests still pass.
- In Storybook confirm the component functions as expected and docs tab is loading properly.
- In the editor, test a gradient picker (e.g. on a background color) to ensure it functions as expected.

## Notes:
- There were some interesting challenges with the `DefinitelyTyped` values for the `gradientParser` library. This led to what felt like pretty liberal use of type assertions. Many of these were to avoid conflicts between union members. Anyway, I know there's a bunch of them and I'm very open to feedback them... so anywhere you see me using an `as`, please take it with a grain of salt and don't hesitate to redirect me if you see a better option.
- I've left inline comments where I felt they'd help clarify what I was doing/thinking at the time, but this one felt like it had more moving parts than I'm used to, so let me know if anything seems off.

chad1008 committed:
Remove from ts-config exclude, rename files to .ts[x]

chad1008 committed:
Add types.ts

chad1008 committed:
Add initial component types

chad1008 committed:
install `@types/gradient-parser`

chad1008 committed:
Add typing to `getGradientAstWithDefault`

chad1008 committed:
Assert that `colorStop.length` is no longer null, due to the `hasUnsupportedValue` check that has already run'

chad1008 committed:
Add typing to `GradientAnglePicker`

chad1008 committed:
Add typing for `gradientTypePicker`

chad1008 committed:
Introduce `ControlPoint` type and apply to `CustomGradientPicker.onChange`'

chad1008 committed:
Add typing to main component utils

chad1008 committed:
Add serializer typing

chad1008 committed:
Fix union conflict on `getGradientAstWithControlPoints`

chad1008 committed:
Add `CustomGradientBar` subcomponent typing

chad1008 committed:
Add typing to `customGradientBarReducer`

chad1008 committed:
Add typing for `GradientBar` utils

chad1008 committed:
Add typing to `ControlPoints`

chad1008 committed:
Refactor `ControlPoints` return value to a single JSX element

chad1008 committed:
Update tests to align with `gradientParser` typing

chad1008 committed:
Update util tests to match typing

chad1008 committed:
Update Storybook

chad1008 committed:
Self-review notes

chad1008 commented on packages/components/src/custom-gradient-picker/gradient-bar/index.tsx:
```diff
@@ -74,19 +86,27 @@ export default function CustomGradientBar( {
 		);
 
+		// NTS: `getHorizontalRelativeGradientPosition` will return `undefined` if the `containerElement` arg is falsey.
+		if ( insertPosition === undefined ) {
+			return;
+		}
```
`getHorizontalRelativeGradientPosition` will return `undefined` if the `containerElement` arg is falsey

chad1008 commented on packages/components/src/custom-gradient-picker/test/serializer.ts:
```diff
@@ -12,24 +12,31 @@ import {
+			serializeGradientColor( {
+				type: 'rgba',
+				//NTS: For entire file - our tests were using numbers, but `gradientParser` (according to DefinitelyTyped) expects strings. I've updated the tests accordingly. Alternatively we could add some logic to our serializer so it accepts both and converts numbers to strings on the fly.
+				value: [ '1', '2', '3', '0.5' ],
```
Note for this entire file: our tests were using numbers, but `gradientParser` (according to DefinitelyTyped) expects strings. I've updated the tests accordingly. Alternatively we could add some logic to our serializer so it accepts both and converts numbers to strings on the fly.

ciampo replied:
Are these changes (from number to string) completely internal to the component, or do they affect the way consumers of `CustomGradientPicker` interact with it?

chad1008 replied:
Internal. These changes are strictly part of a unit test, and the tests themselves are written specifically internal serialization functions that don't get called by a consumer.

ciampo replied:
If they're internal, we should be OK.

chad1008 commented on packages/components/src/custom-gradient-picker/index.tsx:
```diff
@@ -68,7 +83,8 @@ const GradientTypePicker = ( { gradientAST, hasGradient, onChange } ) => {
 				type: 'linear-gradient',
-			} )
+				// NTS: this isn't very elegant, but the `orientation` props between linear and radial gradients don't overlap (according to DefinitelyTyped) so TypeScript gets confused by this manual change.
+			} as gradientParser.LinearGradientNode )
```
This doesn't feel very elegant, but the `orientation` props between linear and radial gradients don't overlap (according to DefinitelyTyped) so TypeScript gets confused by this manual change.

ciampo replied:
Yup, adding this type cast actually hides the real problem: `HORIZONTAL_GRADIENT_ORIENTATION` does not match the expected type for the `orientation` property.

We'd like it to be of type `AngularNode`, but TS currently infers it as `{ type: string; value: number }`, which poses two issues:

- `type` is not narrow enough
- `value`, according to the `gradient-parser` types, should be a `string`, not a `number`

Furthermore, the way the `orientation` prop is re-assigned to the linear gradient object, makes it impossible for TS to understand what's going on.

<details>
<summary>Here's how we could apply a few tweaks and improve the correctness of the code, while making TS happy</summary>

```diff
diff --git a/packages/components/src/custom-gradient-picker/constants.ts b/packages/components/src/custom-gradient-picker/constants.ts
index 9924ff2b3d..842aea9dcb 100644
--- a/packages/components/src/custom-gradient-picker/constants.ts
+++ b/packages/components/src/custom-gradient-picker/constants.ts
@@ -10,8 +10,8 @@ export const DEFAULT_LINEAR_GRADIENT_ANGLE = 180;
 
 export const HORIZONTAL_GRADIENT_ORIENTATION = {
 	type: 'angular',
-	value: 90,
-};
+	value: '90',
+} as const;
 
 export const GRADIENT_OPTIONS = [
 	{ value: 'linear-gradient', label: __( 'Linear' ) },
diff --git a/packages/components/src/custom-gradient-picker/index.tsx b/packages/components/src/custom-gradient-picker/index.tsx
index 8c9b9c7908..a8ffb86e6f 100644
--- a/packages/components/src/custom-gradient-picker/index.tsx
+++ b/packages/components/src/custom-gradient-picker/index.tsx
@@ -76,14 +76,15 @@ const GradientTypePicker = ( {
 	const { type } = gradientAST;
 
 	const onSetLinearGradient = () => {
+		const { orientation, ...restGradientAST } = gradientAST;
 		onChange(
 			serializeGradient( {
-				...gradientAST,
-				...( gradientAST.orientation
-					? {}
-					: { orientation: HORIZONTAL_GRADIENT_ORIENTATION } ),
+				...restGradientAST,
+				orientation: gradientAST.orientation
+					? undefined
+					: HORIZONTAL_GRADIENT_ORIENTATION,
 				type: 'linear-gradient',
-			} as gradientParser.LinearGradientNode )
+			} )
 		);
 	};
 
```
</details>

If we wanted to be more explicit, we could also type `HORIZONTAL_GRADIENT_ORIENTATION` as an `AngularNode`, instead of using `as const` — up to you!

chad1008 replied:
> `value`, according to the `gradient-parser` types, should be a `string`, not a `number`

Shoot - I remember noticing that and then I must have lost track of it while testing different solutions. Good catch, thank you for highlighting it.

> Here's how we could apply a few tweaks and improve the correctness of the code, while making TS happy

Thanks for the diff, very helpful! A question for my own education... Is the `const { orientation, ...restGradientAST } = gradientAST` destructuring, a functional change or more for readability?

<details>
<summary>I wondered if this would work as well as your diff, or if there's a nuance I didn't see:</summary>

```diff

diff --git a/packages/components/src/custom-gradient-picker/constants.ts b/packages/components/src/custom-gradient-picker/constants.ts
index 9924ff2b3d..842aea9dcb 100644
--- a/packages/components/src/custom-gradient-picker/constants.ts
+++ b/packages/components/src/custom-gradient-picker/constants.ts
@@ -10,8 +10,8 @@ export const DEFAULT_LINEAR_GRADIENT_ANGLE = 180;
 
 export const HORIZONTAL_GRADIENT_ORIENTATION = {
 	type: 'angular',
-	value: 90,
-};
+	value: '90',
+} as const;
 
 export const GRADIENT_OPTIONS = [
 	{ value: 'linear-gradient', label: __( 'Linear' ) },
diff --git a/packages/components/src/custom-gradient-picker/index.tsx b/packages/components/src/custom-gradient-picker/index.tsx
index 8c9b9c7908..e19ac3d55f 100644
--- a/packages/components/src/custom-gradient-picker/index.tsx
+++ b/packages/components/src/custom-gradient-picker/index.tsx
@@ -79,11 +79,11 @@ const GradientTypePicker = ( {
 		onChange(
 			serializeGradient( {
 				...gradientAST,
-				...( gradientAST.orientation
-					? {}
-					: { orientation: HORIZONTAL_GRADIENT_ORIENTATION } ),
+				orientation: gradientAST.orientation
+					? undefined
+					: HORIZONTAL_GRADIENT_ORIENTATION,
 				type: 'linear-gradient',
-			} as gradientParser.LinearGradientNode )
+			} )
 		);
 	};
 
```
</details>

ciampo replied:
It was mostly for readability. Your proposed solution works well too and produces a smaller diff 🙌

chad1008 commented on packages/components/src/custom-gradient-picker/index.tsx:
```diff
@@ -82,11 +98,13 @@ const GradientTypePicker = ( { gradientAST, hasGradient, onChange } ) => {
-		if ( next === 'linear-gradient' ) {
+	const handleOnChange = ( next: string | string[] ) => {
+		// NTS: Because `SelectControl.onChange` (below) can accept an array, we need to accept one here. We shouldn't ever see an array here, but I'm adding a check just in case.
+		const nextValue = Array.isArray( next ) ? next[ 0 ] : next;
```
Because `SelectControl.onChange` (below) can accept an array, we need to accept one here. I don't think we should ever actually see an array here, but I'm adding a check just in case.

ciampo replied:
I recently improved types for `SelectControl` in #47390. Can you try to rebase and see if these changes are still necessary?

chad1008 replied:
Very nice, those changes made this much cleaner. Updated in cf33a070b4

chad1008 commented on packages/components/src/custom-gradient-picker/index.tsx:
```diff
@@ -100,30 +118,58 @@ const GradientTypePicker = ( { gradientAST, hasGradient, onChange } ) => {
 			size="__unstable-large"
-			value={ hasGradient && type }
+			// NTS: Small runtime change. Previously, if `type` was undefined (which I don't think should ever happen) this check would return `false`, which would be an invalid value.
+			value={ hasGradient ? type : undefined }
```
Small runtime change. Previously, if `type` was undefined (which I don't think should ever happen) this check would return `false`, which would be an invalid value.

chad1008 commented on packages/components/src/custom-gradient-picker/index.tsx:
```diff
@@ -100,30 +118,58 @@ const GradientTypePicker = ( { gradientAST, hasGradient, onChange } ) => {
+		//NTS: Is there may be a more elegant way to help TS see that for itself? 🤔
+		return {
+			color: getStopCssColor( colorStop ),
+			position: parseInt( colorStop.length!.value ),
```
Is there a more elegant way to help TS see for itself that `colorStop.length` is no longer undefined, and avoid using that `!`? 🤔

ciampo replied:
Mhh, a couple of alternatives that come to mind:

1. Add an explicit `colorStop.length` check where it makes the most sense. The benefit compared to using the `!.` operator is that we would ensure that the length is correct also at runtime. The downside is that we'd introduce a runtime change, which is never great
2. Another approach could be to modify the return type of the `getGradientAstWithDefault` function. Basically, we could tell TS that the return type is a tweaked version of `gradientParser.GradientNode`, where we override the type of `colorStop.length` and remove the possibility that it is `undefined`. Since `gradientParser.GradientNode` is a union type, you'll probably need to use something similar to the `DistributiveOmit` type to preserve the type union.

I would probably discard option 1, and look into option 2 with a time limit. If you can't work it out in a certain amount of time, we can leave the code as-is and improve it another time.

chad1008 replied:
Good idea. I took a stab at the second option of tweaking the return value. It meant using an `Exclude` to remove `undefined` as an option, and then a couple of the `DistributiveOmits` to selectively replace parent props as we work back up the tree.

<details>

<summary>Here's a diff of the proposed changes:</summary>

```diff

diff --git a/packages/components/src/custom-gradient-picker/index.tsx b/packages/components/src/custom-gradient-picker/index.tsx
index ada7b7392c..fda0df040a 100644
--- a/packages/components/src/custom-gradient-picker/index.tsx
+++ b/packages/components/src/custom-gradient-picker/index.tsx
@@ -162,7 +162,7 @@ export function CustomGradientPicker( {
 		// TypeScript doesn't know that `colorStop.length` is not undefined here.
 		return {
 			color: getStopCssColor( colorStop ),
-			position: parseInt( colorStop.length!.value ),
+			position: parseInt( colorStop.length.value ),
 		};
 	} );
 
diff --git a/packages/components/src/custom-gradient-picker/types.ts b/packages/components/src/custom-gradient-picker/types.ts
index 68d026eb7f..cc1e9fc105 100644
--- a/packages/components/src/custom-gradient-picker/types.ts
+++ b/packages/components/src/custom-gradient-picker/types.ts
@@ -125,3 +125,24 @@ export type InsertPointProps = {
 	disableAlpha: boolean;
 	__experimentalIsRenderedInSidebar: boolean;
 };
+
+// The following types are used to define a modified version of the
+// `gradientParser` package's `GradientNode` type. In this version,
+// the `length` property of the `ColorStops` contained in the node cannot be
+// `undefined`. The `getGradientAstWithDefault` and `hasUnsupportedLength` utils
+// ensure this is the case, but TypeScript couldn't detect the narrowing without
+// a more explicit type.
+type DefinedColorStopLength = Exclude<
+	gradientParser.ColorStop[ 'length' ],
+	undefined
+>;
+
+type ColorStopWithDefinedLength = DistributiveOmit<
+	gradientParser.ColorStop,
+	'length'
+> & { length: DefinedColorStopLength };
+
+export type GradientNodeWithDefinedStopLengths = DistributiveOmit<
+	gradientParser.GradientNode,
+	'colorStops'
+> & { colorStops: ColorStopWithDefinedLength[] };
diff --git a/packages/components/src/custom-gradient-picker/utils.ts b/packages/components/src/custom-gradient-picker/utils.ts
index 5b21f94c3e..b355949dc6 100644
--- a/packages/components/src/custom-gradient-picker/utils.ts
+++ b/packages/components/src/custom-gradient-picker/utils.ts
@@ -14,7 +14,7 @@ import {
 	DIRECTIONAL_ORIENTATION_ANGLE_MAP,
 } from './constants';
 import { serializeGradient } from './serializer';
-import type { ControlPoint } from './types';
+import type { ControlPoint, GradientNodeWithDefinedStopLengths } from './types';
 
 extend( [ namesPlugin ] );
 
@@ -75,7 +75,10 @@ export function getGradientAstWithDefault( value?: string ) {
 		gradientAstValue = serializeGradient( gradientAST );
 	}
 
-	return { gradientAST, gradientAstValue };
+	return { gradientAST, gradientAstValue } as {
+		gradientAST: GradientNodeWithDefinedStopLengths;
+		gradientAstValue: string;
+	};
 }
 
 export function getGradientAstWithControlPoints(

```

</details>

If it helps, try messing with this mock `GradientNode`:

```
const test: GradientNodeWithDefinedStopLengths = {
	type: 'linear-gradient',
	colorStops: [
		{
			type: 'literal',
			value: 'cross your fingers...',
			length: { type: '%', value: 'here we go...' },
		},
	],
};
```

With the above diff applied, if you play around with the different `ColorStop` types (`literal`, `hex`, `rgb`, or `rgba`), TypeScript should still expect the right `value` type for that `ColorStop` type, but it will never accept an `undefined` length. Thoughts? 😄

ciampo replied:
Although this exploration is interesting and works as expected, my inclination would be to either keep the original `!` syntax, or to add a `@ts-expect-error` comment with the reason why.

chad1008 replied:
Added a `@ts-expect-error` ✅

chad1008 commented on packages/components/src/custom-gradient-picker/utils.ts:
```diff
@@ -0,0 +1,121 @@
+		},
+		colorStops: gradientAST.colorStops,
+		// NTS: As noted elsewhere, gradientParser's DefinetelyTyped values provide no overlap between the different `orientation` values for linear vs radial nodes.
+	} as gradientParser.LinearGradientNode );
```
As noted elsewhere, gradientParser's `DefinetelyTyped` values provide no overlap between the different `orientation` values for linear vs radial nodes.

ciampo replied:
It looks like this type cast is unnecessary (at least, removing it on my machine doesn't upset TS)

chad1008 committed:
Comment/TODO cleanup

chad1008 committed:
update CHANGELOG

chad1008 commented on packages/components/src/custom-gradient-picker/serializer.ts:
```diff
@@ -0,0 +1,83 @@
+) {
+	if (
+		// NTS: Because we're narrowing to `orientation.type !== 'angular'`, ruling out arrays should have no runtime impact
+		Array.isArray( orientation ) ||
```
Because we're narrowing to `orientation.type !== 'angular'`, ruling out arrays should have no runtime impact

ciampo replied:
Note to future Marco: take a deeper look at this change, make sure it has no runtime impacts

ciampo replied:
I spent some time investigating how the component is currently used on `trunk`, and I can confirm that we're currently never passing an array to this function.

Therefore, your change here is correct — it is a runtime change, but it doesn't seem to cause any regressions. Additionally, it prevents the function from returning an incorrectly formatted value (`undefineddeg`)

chad1008 committed:
TODO removal

chad1008 committed:
partial unit test fixes

ciampo committed:
Use jest mocks for getBoundingClientRect

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx:
```diff
@@ -73,9 +87,9 @@ function GradientColorPickerDropdown( {
 	className,
 	...props
-} ) {
+}: WordPressComponentProps< CustomColorPickerDropdownProps, 'div', true > ) {
```
```suggestion
}: CustomColorPickerDropdownProps ) {
```

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx:
```diff
@@ -73,9 +87,9 @@ function GradientColorPickerDropdown( {
+}: WordPressComponentProps< CustomColorPickerDropdownProps, 'div', true > ) {
 	// Open the popover below the gradient control/insertion point
-	const popoverProps = useMemo(
+	const popoverProps: DropdownProps[ 'popoverProps' ] = useMemo(
```
We could use `as const` to narrow down the type, and therefore avoid the need of using `DropdownProps`

```tsx
const popoverProps = useMemo(
		() =>
			( {
				placement: 'bottom',
				offset: 8,
			} as const ),
		[]
	);
```

chad1008 replied:
👍 0f56c0b5a9

ciampo commented on packages/components/src/custom-gradient-picker/types.ts:
```diff
@@ -0,0 +1,127 @@
+	gradientPickerDomRef: React.RefObject< HTMLDivElement >;
+	ignoreMarkerPosition?: number;
+	value: ControlPoint[];
+	onChange: ( conrolPoints: ControlPoint[] ) => void;
```
```suggestion
	onChange: ( controlPoints: ControlPoint[] ) => void;
```

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx:
```diff
@@ -108,16 +122,28 @@ function ControlPoints( {
-} ) {
-	const controlPointMoveState = useRef();
+}: ControlPointsProps ) {
+	const controlPointMoveState: React.MutableRefObject<
+		ControlPointMoveState | undefined
+	> = useRef();
```
```suggestion
	const controlPointMoveState = useRef< ControlPointMoveState >();
```

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx:
```diff
@@ -108,16 +122,28 @@ function ControlPoints( {
+
 		const { initialPosition, index, significantMoveHappened } =
-			controlPointMoveState.current;
+			controlPointMoveState?.current ?? {};
```
We are guaranteed that `controlPointMoveState` will always be defined

```suggestion
			controlPointMoveState.current ?? {};
```

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx:
```diff
@@ -108,16 +122,28 @@ function ControlPoints( {
+	> = useRef();
 
-	const onMouseMove = ( event ) => {
+	const onMouseMove = ( event: MouseEvent ) => {
```
From our pairing review:

```diff
diff --git a/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx b/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
index 3fddb545ba..e82f5ec79a 100644
--- a/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
+++ b/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
@@ -128,22 +128,28 @@ function ControlPoints( {
 	> = useRef();
 
 	const onMouseMove = ( event: MouseEvent ) => {
-		const relativePosition = getHorizontalRelativeGradientPosition(
-			event.clientX,
-			gradientPickerDomRef.current
-		);
-
-		const { initialPosition, index, significantMoveHappened } =
-			controlPointMoveState?.current ?? {};
+		// TODO: make sure that the "gradientPickerDomRef.current === null" makes sense
+		// ie. if gradientPickerDomRef.current === null, relativePosition would be
+		// `undefined`, which means that `updateControlPointPosition` would end up creating
+		// control points with an `undefined` position. We need to make sure that that's
+		// not a desirable state.
 		if (
 			controlPointMoveState.current === undefined ||
-			initialPosition === undefined ||
-			index === undefined ||
-			relativePosition === undefined
+			gradientPickerDomRef.current === null
 		) {
 			return;
 		}
 
+		// TODO: type getHorizontalRelativeGradientPosition so that it knows that
+		// its return value is `undefined` only if the second argument is null
+		const relativePosition = getHorizontalRelativeGradientPosition(
+			event.clientX,
+			gradientPickerDomRef.current
+		) as number;
+
+		const { initialPosition, index, significantMoveHappened } =
+			controlPointMoveState.current;
+
 		if (
 			! significantMoveHappened &&
 			Math.abs( initialPosition - relativePosition ) >=

```

chad1008 replied:
> make sure that the "gradientPickerDomRef.current === null" makes sense

I did some digging and it looks like the `gradientPickerDomRef.current === null` check is the right move to me - I can't see a case where control points with `undefined` position is desirable 👍 

> type getHorizontalRelativeGradientPosition so that it knows that its return value is `undefined` only if the second argument is null

I tried using a conditional type here, but couldn't get it working so that TypeScript could infer the correct return type.
<details>

<summary>I've experimented with an override instead:</summary>

```diff

diff --git a/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx b/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
index 9668022c62..2f07b6e3b4 100644
--- a/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
+++ b/packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx
@@ -136,7 +136,7 @@ function ControlPoints( {
 		const relativePosition = getHorizontalRelativeGradientPosition(
 			event.clientX,
 			gradientPickerDomRef.current
-		) as number;
+		);
 
 		const { initialPosition, index, significantMoveHappened } =
 			controlPointMoveState.current;
diff --git a/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts b/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
index 271d4fbc16..f82aaac4f0 100644
--- a/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
+++ b/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
@@ -93,6 +93,17 @@ export function updateControlPointColorByPosition(
 	return updateControlPointColor( points, index, newColor );
 }
 
+// `getHorizontalRelativeGradientPosition` should only return `undefined` if
+// the `containerElement` argument is `null` (i.e. when the relevant ref is
+// first created, but before the element is rendered and assigned to the ref).
+export function getHorizontalRelativeGradientPosition(
+	mouseXcoordinate: number,
+	containerElement: HTMLDivElement
+): number;
+export function getHorizontalRelativeGradientPosition(
+	mouseXcoordinate: number,
+	containerElement: null
+): undefined;
 export function getHorizontalRelativeGradientPosition(
 	mouseXCoordinate: number,
 	containerElement: HTMLDivElement | null


```

</details>

But I'm not sure if that's the right way to go.

ciampo replied:
Using overload signatures looks good to me, let's go with your proposed solution :)

chad1008 replied:
Done!

chad1008 replied:
Additional note - when we looked at this change earlier, you floated the idea of a third overload that would accept and return a union of the two possible types, but we didn't feel it was needed. It turns out it is, otherwise we get an error [when inserting a control point](https://github.com/WordPress/gutenberg/blob/9e2185782f1f913a5f9b49c68fab9312d3cf3bb0/packages/components/src/custom-gradient-picker/gradient-bar/index.tsx#L101).

Apparently if the call could accept a union, but none of the overrides do, it errors... even if each union member matches an available override. I did a little hunting and [it looks like this is intentional](https://github.com/WordPress/gutenberg/blob/9e2185782f1f913a5f9b49c68fab9312d3cf3bb0/packages/components/src/custom-gradient-picker/gradient-bar/index.tsx#L101), so I've added the third override for the union.

ciampo commented on packages/components/src/custom-gradient-picker/types.ts:
```diff
@@ -0,0 +1,127 @@
+
+export type ControlPoint = { color: string; position: number };
+
+type DistributiveOmit< T, K extends keyof any > = T extends any
```
This is nice! Let's add a code comment to give more context to why we can't use `Omit` directly

ciampo replied:
https://stackoverflow.com/questions/57103834/typescript-omit-a-property-from-all-interfaces-in-a-union-but-keep-the-union-s

chad1008 committed:
Clean up typing on `useRef` calls

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/control-points.tsx:
```diff
@@ -148,142 +172,158 @@ function ControlPoints( {
 	// This memoization would prevent the event listeners from being properly cleaned.
 	// Instead, we'll pass a ref to the function in our `useEffect` so `cleanEventListeners` itself is no longer a dependency.
-	const cleanEventListenersRef = useRef();
+	const cleanEventListenersRef = useRef< ( () => void ) | undefined >();
```
Small nit: the `| undefined` type is redundant, TypeScript would be able to infer that on its own

```suggestion
	const cleanEventListenersRef = useRef< () => void >();
```

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/utils.ts:
```diff
@@ -0,0 +1,108 @@
+import { MINIMUM_DISTANCE_BETWEEN_POINTS } from './constants';
+import type { ControlPoint } from '../types';
+
+export function clampPercent( value: number ) {
```
We should probably keep the JSDoc comments from the original JS code for all the functions in this file, while removing the type info (which is now specified in TS)

chad1008 replied:
✅ a33606ef4f

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/utils.ts:
```diff
@@ -0,0 +1,108 @@
+export function addControlPoint(
+	points: ControlPoint[],
+	position: number,
+	color: string
```
In the original code, the `color` argument was typed as an object with the `r`, `g`, `b` and `a` properties. Interesting that this type was not correct, and that we we passing a `string` instead

ciampo replied:
Also, following the same convention used below for the `position` prop, we could rewrite this to be

```
	position: ControlPoint[ 'position' ]
	color: ControlPoint[ 'color' ]
```

chad1008 replied:
Yes, I thought that disconnect on the typing was odd as well... there are one or two others in the file as well that were actually strings despite referring back to that `Color` object.

Good point on the convention consistency, updating that and the other instances mentioned below in 7a171c6aa5

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/utils.ts:
```diff
@@ -0,0 +1,108 @@
+export function updateControlPointColor(
+	points: ControlPoint[],
+	index: number,
+	newColor: string
```
Same as above

```suggestion
	newColor: ControlPoint[ 'color' ]
```

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/utils.ts:
```diff
@@ -0,0 +1,108 @@
+export function updateControlPointColorByPosition(
+	points: ControlPoint[],
+	position: ControlPoint[ 'position' ],
+	newColor: string
```
Same as above

```suggestion
	newColor: ControlPoint[ 'color' ]
```

ciampo commented on packages/components/src/custom-gradient-picker/types.ts:
```diff
@@ -0,0 +1,127 @@
+type DistributiveOmit< T, K extends keyof any > = T extends any
+	? Omit< T, K >
+	: never;
+export type ColorStopTypeAndValue = DistributiveOmit<
```
A bit of a nitpick, but the name of the type is a bit misleading. What if the `ColorStop` type had more than 3 props, or if it updated during a release and added/removed props? 

I would either:

- call this time `ColorStopWithoutLength` (simple, but less future-proof)
- keep the `ColorStopTypeAndValue` name, but instead of `Omit`ting the `length` property, I'd explicitly `Pick` only the `type` and `value` props

What do you think?

chad1008 replied:
Never liking to do things the easy way, I think going with a `Pick` approach is best. That said, `Pick` suffers the same "squashing" problem as `Omit` on a union of objects, so we need a `DistributivePick` to make this work. Yes, the geek in me is excited.

I've implemented that in 05b4a18464

Note: I've left `DistributiveOmit` in place, even though it technically isn't used, as it will get used if we apply [some changes](https://github.com/WordPress/gutenberg/pull/48929#discussion_r1134305784) I've proposed elsewhere. Can pull it later if we don't end up needing it.

ciampo replied:
Sounds good. Let's keep this convo open as a reminder to check before merging

chad1008 replied:
Leaving open for now in case we need it, but I went ahead and removed `DistributiveOmit`. Since we didn't keep the changes I mentioned above, the type was no longer being used.

ciampo commented on packages/components/src/custom-gradient-picker/serializer.ts:
```diff
@@ -0,0 +1,82 @@
+			) => {
+				return colorStop?.length?.value === undefined
+					? 0
+					: parseInt( colorStop?.length?.value );
```
The optional chaining here shouldn't be necessary, because of the `colorStop?.length?.value === undefined` check.

Also, for added safety, I wondered if we should check for `NaN` when calling `parseInt`, but then I realised that the result of `getNumericStopValue` is used as the return function of `array.sort`, and I verified that `array.sort(() => NaN)` actually runs corrently, and doesn't change the order of the array elements.

ciampo commented on packages/components/src/custom-gradient-picker/stories/index.tsx:
```diff
@@ -8,16 +12,20 @@ import { useState } from '@wordpress/element';
+const CustomGradientPickerWithState: ComponentStory<
+	typeof CustomGradientPicker
+> = ( props ) => {
+	const [ gradient, setGradient ] = useState< string | undefined >();
```
Nit: We can probably avoid adding the `undefined` type, TS should be able to infer it

ciampo commented on packages/components/src/custom-gradient-picker/types.ts:
```diff
@@ -0,0 +1,127 @@
+	__nextHasNoMargin?: boolean;
+	/**
+	 * The current value of the gradient. Pass a css gradient string (See default value for example).
+	 * Optionally pass in a `null` value to specify no gradient is currently selected.
```
The optional `null` value is not reflected in the type (`string | undefined`)

chad1008 replied:
Good catch, thank you!

Updated in 709aad9ba9. The `null` will hit [the same nullish coalescing operator as an `undefined` value](https://github.com/WordPress/gutenberg/blob/709aad9ba92456affd9b9e2868902d661c83071c/packages/components/src/custom-gradient-picker/utils.ts#L44), so this shouldn't have any side effects.

ciampo commented on packages/components/src/custom-gradient-picker/types.ts:
```diff
@@ -0,0 +1,127 @@
+	 * The function called when a new gradient has been defined. It is passed to
+	 * the `currentGradient` as an arugment.
+	 */
+	onChange: ( currentGradient: string | undefined ) => void;
```
Nit: shorter type

```suggestion
	onChange: ( currentGradient?: string ) => void;
```

ciampo reviewed (commented):
Great job so far, @chad1008 !

This is not an easy one

chad1008 committed:
fix `GradienttypePicker` `onSetLinearGradient`.

chad1008 committed:
Improve `GradientTypePicker`s `handleOnChange` type handling to align with updates to `SelectCotnrol`

chad1008 committed:
Remove unneeded `WordPressComponentProps`

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chad1008 committed:
`as const` instead of importing prop types from `Dropdown`

chad1008 committed:
Typo

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chad1008 committed:
Remove unneeded `undefined` typing

chad1008 committed:
Improve `ControlPoints` typeguarding logic

chad1008 committed:
Add comment describing `DistributiveOmit`

chad1008 committed:
Remove unneeded `undefined` typing

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chad1008 committed:
Restore original gradient bar util JSDoc comments, minus type info

chad1008 committed:
In utils, type `color` props as `ControlPoint[color]`

chad1008 committed:
Convert `ColorStopTypeAndValue` to a `Pick`-based approach

chad1008 committed:
remove extra optional chaining

chad1008 committed:
remove unneeded `undefined` typing

chad1008 committed:
include option `null` type for `CustomGradientPickerProps.value`

chad1008 committed:
Code cleanup

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chad1008 commented:
Whew, thanks for all of the detailed feedback, as always @ciampo.

I think I've addressed everything. There are a couple of unresolved conversations in the thread I had questions or needed feedback on.

Beyond that, the conflicts listed are updating the CHANGELOG and addressing some added `@ts-nocheck` comments that were added in #47764. Once we've resolved other open feedback I'll rebase those in and remove them.

ciampo commented on packages/components/src/custom-gradient-picker/types.ts:
```diff
@@ -0,0 +1,140 @@
+	value?: string | null;
+	/**
+	 * The function called when a new gradient has been defined. It is passed to
+	 * the `currentGradient` as an arugment.
```
```suggestion
	 * the `currentGradient` as an argument.
```

ciampo reviewed (commented):
Let's rebase to incorporate changes, and then I'll take another look!

chad1008 committed:
fix typos

chad1008 committed:
replace `!` with `@ts-expect-error` in `CustomGradientPicker` `controlPoints` declaration

chad1008 committed:
Apply overrides to clarify `getHorizontalRelativeGradientPosition` return type

chad1008 committed:
remove `DistributiveOmit` in favor of `DistributivePick`

chad1008 committed:
Remove `@ts-nocheck`s

chad1008 committed:
Mark `Gradient.gradient` from `PaletteEdit` as optional

chad1008 commented:
Rebased and incorporated changes. I did need to make one small change (7595d97dcc2569332eed97b6b2a99b1677bc2bd4) to the new types in `PaletteEdit`, which treated the gradient string as a required value, when it's actually optional. `CustomGradientPicker` will apply a default gradient if one isn't provided.

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/utils.ts:
```diff
@@ -0,0 +1,197 @@
+	value: ControlPoint[],
+	initialIndex: number,
+	newPosition: number,
+	minDistance = MINIMUM_DISTANCE_BETWEEN_POINTS
```
It probably doesn't hurt to specify that `minDistance` is a `number`

```suggestion
	minDistance: number = MINIMUM_DISTANCE_BETWEEN_POINTS
```

ciampo commented on packages/components/src/custom-gradient-picker/gradient-bar/utils.ts:
```diff
@@ -0,0 +1,197 @@
+	mouseXcoordinate: number,
+	containerElement: null
+): undefined;
+export function getHorizontalRelativeGradientPosition(
+	mouseXcoordinate: number,
+	containerElement: HTMLDivElement | null
+): number | undefined;
```
I think that your initial implementation was correct — we don't really need to advertise this type for the `getHorizontalRelativeGradientPosition` function

```suggestion
```

chad1008 replied:
[This comment probably got lost](https://github.com/WordPress/gutenberg/pull/48929#discussion_r1137752782) in the mountain of threads we've accumulated, but I think we do actually need the union override.

If you don't get the same error, or if there's a better solution, I'm happy to investigate further!

ciampo replied:
I definitely missed it 😓 

Anyway, I took a look and I see why we'd need the third overload. Although I also looked at the TS error, and I think that a better solution would be to tweak the early return check in the `GradientBar`, which would then allows us to remove that third function overload:

```diff
diff --git a/packages/components/src/custom-gradient-picker/gradient-bar/index.tsx b/packages/components/src/custom-gradient-picker/gradient-bar/index.tsx
index 526ee53bff..225a5fb757 100644
--- a/packages/components/src/custom-gradient-picker/gradient-bar/index.tsx
+++ b/packages/components/src/custom-gradient-picker/gradient-bar/index.tsx
@@ -96,15 +96,15 @@ export default function CustomGradientBar( {
 	const onMouseEnterAndMove: MouseEventHandler< HTMLDivElement > = (
 		event
 	) => {
+		if ( ! gradientMarkersContainerDomRef.current ) {
+			return;
+		}
+
 		const insertPosition = getHorizontalRelativeGradientPosition(
 			event.clientX,
 			gradientMarkersContainerDomRef.current
 		);
 
-		if ( insertPosition === undefined ) {
-			return;
-		}
-
 		// If the insert point is close to an existing control point don't show it.
 		if (
 			controlPoints.some( ( { position } ) => {
diff --git a/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts b/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
index 6e9417328c..88daf2b2c1 100644
--- a/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
+++ b/packages/components/src/custom-gradient-picker/gradient-bar/utils.ts
@@ -178,10 +178,6 @@ export function getHorizontalRelativeGradientPosition(
 	mouseXcoordinate: number,
 	containerElement: null
 ): undefined;
-export function getHorizontalRelativeGradientPosition(
-	mouseXcoordinate: number,
-	containerElement: HTMLDivElement | null
-): number | undefined;
 export function getHorizontalRelativeGradientPosition(
 	mouseXCoordinate: number,
 	containerElement: HTMLDivElement | null
```

What do you think?

chad1008 replied:
Love the change.

Hate that I didn't think of it myself.

Implemented: 465a59d285

ciampo commented on packages/components/src/custom-gradient-picker/utils.ts:
```diff
@@ -16,60 +14,73 @@ import {
+		orientation: {
+			...HORIZONTAL_GRADIENT_ORIENTATION,
+			value: `${ HORIZONTAL_GRADIENT_ORIENTATION.value }`,
+		},
```
I believe we can undo this change, since we change the `value` property to be a string directly where `HORIZONTAL_GRADIENT_ORIENTATION` is declared

ciampo commented on packages/components/src/palette-edit/types.ts:
```diff
@@ -16,7 +16,7 @@ export type Color = {
 
 export type Gradient = {
-	gradient: string;
+	gradient?: string;
```
I think that a better change would be to change the type of `CustomGradientPicker`'s `onChange` function, since from a quick look it seems that it can never have `undefined` as its argument:

```diff
diff --git a/packages/components/src/custom-gradient-picker/types.ts b/packages/components/src/custom-gradient-picker/types.ts
index 3d9d570382..9e241dd043 100644
--- a/packages/components/src/custom-gradient-picker/types.ts
+++ b/packages/components/src/custom-gradient-picker/types.ts
@@ -23,7 +23,7 @@ export type CustomGradientPickerProps = {
 	 * The function called when a new gradient has been defined. It is passed to
 	 * the `currentGradient` as an argument.
 	 */
-	onChange: ( currentGradient?: string ) => void;
+	onChange: ( currentGradient: string ) => void;
 	/**
 	 * Whether this is rendered in the sidebar.
 	 *
diff --git a/packages/components/src/palette-edit/index.tsx b/packages/components/src/palette-edit/index.tsx
index 7882950431..9b54c5ed0c 100644
--- a/packages/components/src/palette-edit/index.tsx
+++ b/packages/components/src/palette-edit/index.tsx
@@ -133,7 +133,7 @@ function ColorPickerPopover< T extends Color | Gradient >( {
 						__nextHasNoMargin
 						__experimentalIsRenderedInSidebar
 						value={ element.gradient }
-						onChange={ ( newGradient: Gradient[ 'gradient' ] ) => {
+						onChange={ ( newGradient ) => {
 							onChange( {
 								...element,
 								gradient: newGradient,

```

On a separate note, `PaletteEdit` types are quite convoluted due to the `Color | Gradient` union — I had a time-boxed go to see if I could simplify them further, but didn't manage to get to a good point on time

chad1008 replied:
Sounds good to me! ef06890dee

ciampo commented on packages/components/src/custom-gradient-picker/stories/index.tsx:
```diff
@@ -8,16 +12,20 @@ import { useState } from '@wordpress/element';
 	component: CustomGradientPicker,
-	argTypes: {
-		__nextHasNoMargin: { control: { type: 'boolean' } },
+	parameters: {
```
Let's add support for logging callbacks directly in Storybook by passing the `actions` property here, and by forwarding the `onChange` prop in the story code

```diff
diff --git a/packages/components/src/custom-gradient-picker/stories/index.tsx b/packages/components/src/custom-gradient-picker/stories/index.tsx
index 98c50b778b..3648db3abe 100644
--- a/packages/components/src/custom-gradient-picker/stories/index.tsx
+++ b/packages/components/src/custom-gradient-picker/stories/index.tsx
@@ -16,6 +16,7 @@ const meta: ComponentMeta< typeof CustomGradientPicker > = {
 	title: 'Components/CustomGradientPicker',
 	component: CustomGradientPicker,
 	parameters: {
+		actions: { argTypesRegex: '^on.*' },
 		controls: { expanded: true },
 		docs: { source: { state: 'open' } },
 	},
@@ -24,13 +25,16 @@ export default meta;
 
 const CustomGradientPickerWithState: ComponentStory<
 	typeof CustomGradientPicker
-> = ( props ) => {
+> = ( { onChange, ...props } ) => {
 	const [ gradient, setGradient ] = useState< string >();
 	return (
 		<CustomGradientPicker
 			{ ...props }
 			value={ gradient }
-			onChange={ setGradient }
+			onChange={ ( newGradient ) => {
+				setGradient( newGradient );
+				onChange( newGradient );
+			} }
 		/>
 	);
 };

```

chad1008 replied:
✅ d0f1d81195

chad1008 committed:
Mark `CustomGradientBar.__experimentalIsRenderedInSidebar` as optional with `false` default

chad1008 committed:
Type `minDistance` as `number`

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chad1008 committed:
Revert string casting that is no longer needed

chad1008 committed:
Remove unneeded type assertion

chad1008 committed:
Require a value for the input of `CustomGradientPickerProps.onChange` to better align with `ColorPickerPopover` types

chad1008 committed:
Add support for Storybook callback logging

chad1008 commented:
Applied most recent feedback. I believe there is only [one remaining open item](https://github.com/WordPress/gutenberg/pull/48929#discussion_r1139017342).

I did introduce one additional [small change](https://github.com/WordPress/gutenberg/pull/48929/commits/8f9e1887b43b5333daab9d336422d05b8a0df081). tl;dr: `CustomGradientBar` was treating `__experimentalIsRenderedInSidebar` as required, when it shouldn't be, and isn't provided by consumers like `DuotonPicker` so it's now optional with a default of `false`, which aligns with the main `CustomGradientPicker` component.

ciampo reviewed (approved):
This was a tough one! Thank you @chad1008 for the patience in addressing all the feedback. I think we did a good job and introduced very few runtime changes 🚀 

Good to merge once [this last comment](https://github.com/WordPress/gutenberg/pull/48929#discussion_r1140305322) is addressed

chad1008 committed:
Apply Marco's brilliance

chad1008 commented:
Last change implemented, waiting on CI before merging...
